File: server/services/game-export/unity/templates/scripts/systems/QuestSystem.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using Insimul.Data;

namespace Insimul.Systems
{
    /// <summary>
    /// Quest objective data mirroring Insimul's QuestObjective interface.
    /// </summary>
    [Serializable]
    public class QuestObjective
    {
        public string id;
        public string questId;
        public string type;
        public string description;
        public bool completed;
        public int requiredCount = 1;
        public int currentCount;

        /// <summary>Time limit in seconds (0 = untimed).</summary>
        public float timeLimitSeconds;

        /// <summary>Game-time when objective was started (-1 = not started).</summary>
        public float startedAt = -1f;

        /// <summary>Number of English hints requested (navigation/directions).</summary>
        public int hintsRequested;

        /// <summary>Whether a GPS-style waypoint is currently shown.</summary>
        public bool showWaypoint;

        /// <summary>Vocabulary category for scavenger hunt rotation.</summary>
        public string vocabularyCategory;

        /// <summary>Target words for vocabulary objectives (empty = any word counts).</summary>
        public List<string> targetWords = new();

        /// <summary>Words already used (for deduplication).</summary>
        public List<string> wordsUsed = new();

        /// <summary>Item name for collect_item / deliver_item objectives.</summary>
        public string itemName;

        /// <summary>NPC ID for NPC-targeted objectives.</summary>
        public string npcId;
    }

    /// <summary>
    /// Result struct returned by TrackCollectedItemByName for each matched objective.
    /// </summary>
    [Serializable]
    public struct CollectedItemMatch
    {
        public string questId;
        public string objectiveId;
        public string matchedName;
        public int collectedCount;
        public int requiredCount;
        public bool completed;
    }

    public class QuestSystem : MonoBehaviour
    {
        public event Action<string, string> OnStoryTTS;
        public event Action<string, string, string> OnQuestItemCollected;

        /// <summary>Scavenger hunt categories for vocabulary rotation.</summary>
        public static readonly string[] SCAVENGER_CATEGORIES = {
            "food", "colors", "animals", "clothing", "household",
            "nature", "body", "professions", "transportation", "weather"
        };

        private List<InsimulQuestData> _allQuests = new();
        private List<string> _activeQuestIds = new();
        private HashSet<string> _completedQuestIds = new();
        private List<QuestObjective> _objectives = new();
        private float _gameTime;

        public int QuestCount => _allQuests.Count;
        public int ActiveCount => _activeQuestIds.Count;

        private void Update()
        {
            _gameTime += Time.deltaTime;
        }

        public void LoadFromData(InsimulWorldIR worldData)
        {
            if (worldData?.systems?.quests == null) return;
            _allQuests.AddRange(worldData.systems.quests);
            Debug.Log($"[Insimul] QuestSystem loaded {_allQuests.Count} quests");
        }

        public bool AcceptQuest(string questId)
        {
            var quest = _allQuests.Find(q => q.id == questId);
            if (quest == null || _activeQuestIds.Contains(questId)) return false;
            _activeQuestIds.Add(questId);
            Debug.Log($"[Insimul] Quest accepted: {quest.title}");
            OnQuestAccepted?.Invoke(questId);
            return true;
        }

        public bool CompleteQuest(string questId)
        {
            if (!_activeQuestIds.Contains(questId)) return false;
            _activeQuestIds.Remove(questId);
            _completedQuestIds.Add(questId);
            Debug.Log($"[Insimul] Quest completed: {questId}");
            OnQuestCompleted?.Invoke(questId);
            return true;
        }

        public bool CompleteObjective(string questId, string objectiveId)
        {
            var obj = _objectives.Find(o => o.questId == questId && o.id == objectiveId);
            if (obj == null || obj.completed) return false;
            obj.completed = true;
            Debug.Log($"[Insimul] Objective completed: {questId}/{objectiveId}");
            return true;
        }

        /// <summary>Check timed objectives and return descriptions of expired ones.</summary>
        public List<string> CheckTimedObjectives()
        {
            var expired = new List<string>();
            foreach (var obj in _objectives)
            {
                if (obj.completed) continue;
                if (obj.timeLimitSeconds <= 0f || obj.startedAt < 0f) continue;

                float elapsed = _gameTime - obj.startedAt;
                if (elapsed > obj.timeLimitSeconds)
                {
                    obj.completed = true;
                    expired.Add($"Time expired: {obj.description}");
                    Debug.Log($"[Insimul] Timed objective expired: {obj.id}");
                }
            }
            return expired;
        }

        /// <summary>Get remaining seconds for a timed objective, or -1 if untimed.</summary>
        public float GetObjectiveTimeRemaining(string objectiveId)
        {
            var obj = _objectives.Find(o => o.id == objectiveId);
            if (obj != null && obj.timeLimitSeconds > 0f && obj.startedAt >= 0f)
            {
                float elapsed = _gameTime - obj.startedAt;
                return Mathf.Max(0f, obj.timeLimitSeconds - elapsed);
            }
            return -1f;
        }

        /// <summary>Request a GPS-style waypoint hint. Returns English hint or null.</summary>
        public string RequestNavigationHint(string questId = null)
        {
            foreach (var obj in _objectives)
            {
                if (obj.completed) continue;
                if (!string.IsNullOrEmpty(questId) && obj.questId != questId) continue;
                if (obj.type != "navigate_language" && obj.type != "follow_directions") continue;

                obj.hintsRequested++;
                obj.showWaypoint = true;
                Debug.Log($"[Insimul] Navigation hint #{obj.hintsRequested} for {obj.id}");
                return obj.description;
            }
            return null;
        }

        /// <summary>Get next scavenger hunt category (round-robin).</summary>
        public static string GetNextScavengerCategory(int lastCategoryIndex)
        {
            int next = (lastCategoryIndex + 1) % SCAVENGER_CATEGORIES.Length;
            return SCAVENGER_CATEGORIES[next];
        }

        /// <summary>Track vocabulary usage for use_vocabulary / collect_vocabulary objectives.</summary>
        public void TrackVocabularyUsage(string word, string questId = null)
        {
            string lowerWord = word.ToLower();

            foreach (var obj in _objectives)
            {
                if (obj.completed) continue;
                if (!string.IsNullOrEmpty(questId) && obj.questId != questId) continue;
                if (obj.type != "use_vocabulary" && obj.type != "collect_vocabulary") continue;

                // If targetWords specified, only count matching words
                if (obj.targetWords.Count > 0 && !obj.targetWords.Contains(lowerWord)) continue;

                // Don't double-count the same word
                if (obj.wordsUsed.Contains(lowerWord)) continue;

                obj.wordsUsed.Add(lowerWord);
                obj.currentCount++;

                if (obj.currentCount >= (obj.requiredCount > 0 ? obj.requiredCount : 10))
                {
                    CompleteObjective(obj.questId, obj.id);
                }
            }
        }

        /// <summary>Track a conversation turn for complete_conversation objectives.</summary>
        public void TrackConversationTurn(string[] keywords = null, string questId = null)
        {
            foreach (var obj in _objectives)
            {
                if (obj.completed) continue;
                if (!string.IsNullOrEmpty(questId) && obj.questId != questId) continue;
                if (obj.type != "complete_conversation") continue;

                // Every conversation turn counts as progress
                obj.currentCount++;

                if (obj.currentCount >= (obj.requiredCount > 0 ? obj.requiredCount : 5))
                {
                    CompleteObjective(obj.questId, obj.id);
                }
            }
        }

        /// <summary>Track a pronunciation attempt for pronunciation_check objectives.</summary>
        /// <param name="score">Pronunciation accuracy score (0-100)</param>
        public void TrackPronunciationAttempt(bool passed, float score = 0f, string questId = null)
        {
            if (!passed) return;

            foreach (var obj in _objectives)
            {
                if (obj.completed) continue;
                if (!string.IsNullOrEmpty(questId) && obj.questId != questId) continue;
                if (obj.type != "pronunciation_check") continue;

                obj.currentCount++;

                if (obj.currentCount >= (obj.requiredCount > 0 ? obj.requiredCount : 3))
                {
                    CompleteObjective(obj.questId, obj.id);
                }
            }
        }

        /// <summary>Track item delivery to an NPC for deliver_item objectives.</summary>
        public void TrackItemDelivery(string npcId, string[] playerItemNames, string questId = null)
        {
            var normalizedItems = new HashSet<string>(
                playerItemNames.Select(n => n.ToLowerInvariant()));

            foreach (var obj in _objectives)
            {
                if (obj.completed) continue;
                if (!string.IsNullOrEmpty(questId) && obj.questId != questId) continue;
                if (obj.type != "deliver_item") continue;
                if (!string.IsNullOrEmpty(obj.npcId) && obj.npcId != npcId) continue;

                string itemLower = (obj.itemName ?? "").ToLowerInvariant();
                if (!string.IsNullOrEmpty(itemLower) && normalizedItems.Contains(itemLower))
                {
                    CompleteObjective(obj.questId, obj.id);
                }
            }
        }

        /// <summary>Track a crafted item for craft_item objectives.</summary>
        public void TrackCraftedItem(string recipeId, string itemName, string stationType = null, string questId = null)
        {
            string lowerItem = itemName.ToLowerInvariant();
            string lowerRecipe = recipeId.ToLowerInvariant();

            foreach (var obj in _objectives)
            {
                if (obj.completed) continue;
                if (!string.IsNullOrEmpty(questId) && obj.questId != questId) continue;
                if (obj.type != "craft_item") continue;

                string objItemLower = (obj.itemName ?? "").ToLowerInvariant();
                bool nameMatch = false;

                // Exact match on item name or recipe id
                if (!string.IsNullOrEmpty(objItemLower) &&
                    (objItemLower == lowerItem || objItemLower == lowerRecipe))
                {
                    nameMatch = true;
                }
                // No item name on objective means any craft counts
                else if (string.IsNullOrEmpty(objItemLower))
                {
                    nameMatch = true;
                }

                if (!nameMatch) continue;

                obj.currentCount++;

                int required = obj.requiredCount > 0 ? obj.requiredCount : 1;
                if (obj.currentCount >= required)
                {
                    CompleteObjective(obj.questId, obj.id);
                }
            }
        }

        /// <summary>Track a gift given to an NPC for give_gift objectives.</summary>
        public void TrackGiftGiven(string npcId, string itemName, string questId = null)
        {
            foreach (var obj in _objectives)
            {
                if (obj.completed) continue;
                if (!string.IsNullOrEmpty(questId) && obj.questId != questId) continue;
                if (obj.type != "give_gift") continue;
                if (!string.IsNullOrEmpty(obj.npcId) && obj.npcId != npcId) continue;

                CompleteObjective(obj.questId, obj.id);
            }
        }

        /// <summary>Track a collected item by name for collect_item / identify_object / find_vocabulary_items objectives.</summary>
        public List<CollectedItemMatch> TrackCollectedItemByName(string itemName, string category = null, string questId = null)
        {
            var matches = new List<CollectedItemMatch>();
            string lowerItem = itemName.ToLowerInvariant();

            foreach (var obj in _objectives)
            {
                if (obj.completed) continue;
                if (!string.IsNullOrEmpty(questId) && obj.questId != questId) continue;
                if (obj.type != "collect_item" && obj.type != "identify_object" && obj.type != "find_vocabulary_items") continue;

                string objItemLower = (obj.itemName ?? "").ToLowerInvariant();
                bool nameMatch = false;

                // Exact match
                if (!string.IsNullOrEmpty(objItemLower) && objItemLower == lowerItem)
                {
                    nameMatch = true;
                }
                // Partial match: item name contains objective name or vice versa
                else if (!string.IsNullOrEmpty(objItemLower) && (lowerItem.Contains(objItemLower) || objItemLower.Contains(lowerItem)))
                {
                    nameMatch = true;
                }
                // Category match: objective's vocabularyCategory matches the provided category
                else if (!string.IsNullOrEmpty(category) && !string.IsNullOrEmpty(obj.vocabularyCategory) &&
                         string.Equals(obj.vocabularyCategory, category, StringComparison.OrdinalIgnoreCase))
                {
                    nameMatch = true;
                }
                // No item name on objective means any item counts
                else if (string.IsNullOrEmpty(objItemLower) && string.IsNullOrEmpty(obj.vocabularyCategory))
                {
                    nameMatch = true;
                }

                if (!nameMatch) continue;

                obj.currentCount++;

                int required = obj.requiredCount > 0 ? obj.requiredCount : 1;
                bool completed = obj.currentCount >= required;
                if (completed)
                {
                    CompleteObjective(obj.questId, obj.id);
                }

                matches.Add(new CollectedItemMatch
                {
                    questId = obj.questId,
                    objectiveId = obj.id,
                    matchedName = itemName,
                    collectedCount = obj.currentCount,
                    requiredCount = required,
                    completed = completed
                });

                OnQuestItemCollected?.Invoke(obj.questId, obj.id, itemName);
            }
            return matches;
        }

        /// <summary>
        /// Register a building-check callback so spawned items avoid building interiors.
        /// The callback receives world-space (x, z) and returns true if inside a building.
        /// </summary>
        public void SetPointInBuildingCheck(Func<float, float, bool> check)
        {
            _pointInBuildingCheck = check;
        }

        /// <summary>Generate spread-out item positions that avoid building interiors.</summary>
        public List<Vector3> GenerateItemPositions(int count)
        {
            var positions = new List<Vector3>();
            float radius = 30f;

            for (int i = 0; i < count; i++)
            {
                float x = 0f, z = 0f;
                for (int attempt = 0; attempt < 8; attempt++)
                {
                    float angle = (Mathf.PI * 2f * i) / count + UnityEngine.Random.Range(0f, 0.5f);
                    float dist = 10f + UnityEngine.Random.Range(0f, radius);
                    x = Mathf.Cos(angle) * dist;
                    z = Mathf.Sin(angle) * dist;
                    if (_pointInBuildingCheck == null || !_pointInBuildingCheck(x, z))
                        break;
                }
                positions.Add(new Vector3(x, 0.5f, z));
            }
            return positions;
        }

        /// <summary>Generate a single location position that avoids building interiors.</summary>
        public Vector3 GenerateLocationPosition()
        {
            for (int attempt = 0; attempt < 8; attempt++)
            {
                float angle = UnityEngine.Random.Range(0f, Mathf.PI * 2f);
                float dist = 20f + UnityEngine.Random.Range(0f, 20f);
                float x = Mathf.Cos(angle) * dist;
                float z = Mathf.Sin(angle) * dist;
                if (_pointInBuildingCheck == null || !_pointInBuildingCheck(x, z))
                    return new Vector3(x, 0f, z);
            }
            // Fallback — push farther out
            float fallbackAngle = UnityEngine.Random.Range(0f, Mathf.PI * 2f);
            float fallbackDist = 40f + UnityEngine.Random.Range(0f, 10f);
            return new Vector3(Mathf.Cos(fallbackAngle) * fallbackDist, 0f, Mathf.Sin(fallbackAngle) * fallbackDist);
        }

        /// <summary>
        /// Attach debug metadata to a quest marker GameObject (used for hover tooltips).
        /// Replaces floating 3D text labels with lightweight metadata.
        /// </summary>
        public static void SetMarkerDebugLabel(GameObject marker, string label)
        {
            if (marker == null) return;
            var meta = marker.GetComponent<QuestMarkerMeta>();
            if (meta == null) meta = marker.AddComponent<QuestMarkerMeta>();
            meta.debugLabel = label;
        }

        public InsimulQuestData GetQuest(string id) => _allQuests.Find(q => q.id == id);
        public List<InsimulQuestData> GetAllQuests() => new List<InsimulQuestData>(_allQuests);
        public List<InsimulQuestData> GetActiveQuests() =>
            _allQuests.FindAll(q => _activeQuestIds.Contains(q.id));
        public List<InsimulQuestData> GetCompletedQuests() =>
            _allQuests.FindAll(q => _completedQuestIds.Contains(q.id));
        public List<InsimulQuestData> GetAvailableQuests() =>
            _allQuests.FindAll(q => !_activeQuestIds.Contains(q.id) && !_completedQuestIds.Contains(q.id));
        public bool IsQuestActive(string id) => _activeQuestIds.Contains(id);
        public bool IsQuestCompleted(string id) => _completedQuestIds.Contains(id);
        public List<QuestObjective> GetObjectivesForQuest(string questId) =>
            _objectives.FindAll(o => o.questId == questId);

        public event Action<string> OnQuestAccepted;
        public event Action<string> OnQuestCompleted;

        private Func<float, float, bool> _pointInBuildingCheck;
    }

    /// <summary>
    /// Lightweight metadata component for quest markers (debug labels, tooltips).
    /// </summary>
    public class QuestMarkerMeta : MonoBehaviour
    {
        public string debugLabel;
    }
}

File: server/services/game-export/unreal/templates/source/systems/QuestSystem.cpp
#include "QuestSystem.h"
#include "Dom/JsonObject.h"
#include "Serialization/JsonReader.h"
#include "Serialization/JsonSerializer.h"

void UQuestSystem::Initialize(FSubsystemCollectionBase& Collection)
{
    Super::Initialize(Collection);
    UE_LOG(LogTemp, Log, TEXT("[Insimul] QuestSystem initialized"));
}

void UQuestSystem::Deinitialize()
{
    Super::Deinitialize();
}

void UQuestSystem::LoadFromIR(const FString& JsonString)
{
    TSharedPtr<FJsonObject> Root;
    TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(JsonString);
    if (!FJsonSerializer::Deserialize(Reader, Root) || !Root.IsValid()) return;

    const TSharedPtr<FJsonObject>* SystemsObj;
    if (Root->TryGetObjectField(TEXT("systems"), SystemsObj))
    {
        const TArray<TSharedPtr<FJsonValue>>* QuestsArr;
        if ((*SystemsObj)->TryGetArrayField(TEXT("quests"), QuestsArr))
        {
            QuestCount = QuestsArr->Num();
            UE_LOG(LogTemp, Log, TEXT("[Insimul] Loaded %d quests"), QuestCount);
        }
    }
}

bool UQuestSystem::AcceptQuest(const FString& QuestId)
{
    UE_LOG(LogTemp, Log, TEXT("[Insimul] AcceptQuest: %s"), *QuestId);
    return true;
}

bool UQuestSystem::CompleteObjective(const FString& QuestId, const FString& ObjectiveId)
{
    for (auto& Obj : Objectives)
    {
        if (Obj.QuestId == QuestId && Obj.Id == ObjectiveId && !Obj.bCompleted)
        {
            Obj.bCompleted = true;
            UE_LOG(LogTemp, Log, TEXT("[Insimul] CompleteObjective: %s / %s"), *QuestId, *ObjectiveId);
            return true;
        }
    }
    return false;
}

TArray<FString> UQuestSystem::CheckTimedObjectives()
{
    TArray<FString> Expired;

    for (auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;
        if (Obj.TimeLimitSeconds <= 0.f || Obj.StartedAt < 0.f) continue;

        float Elapsed = GameTime - Obj.StartedAt;
        if (Elapsed > Obj.TimeLimitSeconds)
        {
            Obj.bCompleted = true;
            Expired.Add(FString::Printf(TEXT("Time expired: %s"), *Obj.Description));
            UE_LOG(LogTemp, Log, TEXT("[Insimul] Timed objective expired: %s"), *Obj.Id);
        }
    }
    return Expired;
}

float UQuestSystem::GetObjectiveTimeRemaining(const FString& ObjectiveId) const
{
    for (const auto& Obj : Objectives)
    {
        if (Obj.Id == ObjectiveId && Obj.TimeLimitSeconds > 0.f && Obj.StartedAt >= 0.f)
        {
            float Elapsed = GameTime - Obj.StartedAt;
            return FMath::Max(0.f, Obj.TimeLimitSeconds - Elapsed);
        }
    }
    return -1.f;
}

FString UQuestSystem::RequestNavigationHint(const FString& QuestId)
{
    for (auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;
        if (!QuestId.IsEmpty() && Obj.QuestId != QuestId) continue;
        if (Obj.Type != TEXT("navigate_language") && Obj.Type != TEXT("follow_directions")) continue;

        Obj.HintsRequested++;
        Obj.bShowWaypoint = true;
        UE_LOG(LogTemp, Log, TEXT("[Insimul] Navigation hint requested for %s (hint #%d)"), *Obj.Id, Obj.HintsRequested);
        return Obj.Description;
    }
    return FString();
}

void UQuestSystem::TrackVocabularyUsage(const FString& Word, const FString& QuestId)
{
    FString LowerWord = Word.ToLower();

    for (auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;
        if (!QuestId.IsEmpty() && Obj.QuestId != QuestId) continue;
        if (Obj.Type != TEXT("use_vocabulary") && Obj.Type != TEXT("collect_vocabulary")) continue;

        // If targetWords specified, only count matching words
        if (Obj.TargetWords.Num() > 0 && !Obj.TargetWords.Contains(LowerWord)) continue;

        // Don't double-count the same word
        if (Obj.WordsUsed.Contains(LowerWord)) continue;

        Obj.WordsUsed.Add(LowerWord);
        Obj.CurrentCount++;

        if (Obj.CurrentCount >= (Obj.RequiredCount > 0 ? Obj.RequiredCount : 10))
        {
            Obj.bCompleted = true;
            UE_LOG(LogTemp, Log, TEXT("[Insimul] Vocabulary objective completed: %s"), *Obj.Id);
        }
    }
}

void UQuestSystem::TrackConversationTurn(const TArray<FString>& Keywords, const FString& QuestId)
{
    // Conversation-only objective types that progress on each conversation turn
    static const TSet<FString> ConversationOnlyTypes = {
        TEXT("complete_conversation"), TEXT("order_food"), TEXT("haggle_price"),
        TEXT("listen_and_repeat"), TEXT("ask_for_directions"), TEXT("describe_scene"),
        TEXT("write_response"), TEXT("build_friendship")
    };

    for (auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;
        if (!QuestId.IsEmpty() && Obj.QuestId != QuestId) continue;
        if (!ConversationOnlyTypes.Contains(Obj.Type)) continue;

        // Every conversation turn counts as progress
        Obj.CurrentCount++;

        if (Obj.CurrentCount >= (Obj.RequiredCount > 0 ? Obj.RequiredCount : 5))
        {
            Obj.bCompleted = true;
            UE_LOG(LogTemp, Log, TEXT("[Insimul] Conversation objective completed: %s"), *Obj.Id);
        }
    }
}

void UQuestSystem::TrackPronunciationAttempt(bool bPassed, float Score, const FString& QuestId)
{
    if (!bPassed) return;

    for (auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;
        if (!QuestId.IsEmpty() && Obj.QuestId != QuestId) continue;
        if (Obj.Type != TEXT("pronunciation_check")) continue;

        Obj.CurrentCount++;

        if (Obj.CurrentCount >= (Obj.RequiredCount > 0 ? Obj.RequiredCount : 3))
        {
            Obj.bCompleted = true;
            UE_LOG(LogTemp, Log, TEXT("[Insimul] Pronunciation objective completed: %s"), *Obj.Id);
        }
    }
}

void UQuestSystem::TrackWritingSubmission(const FString& Text, int32 WordCount, const FString& QuestId)
{
    for (auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;
        if (!QuestId.IsEmpty() && Obj.QuestId != QuestId) continue;
        if (Obj.Type != TEXT("write_response") && Obj.Type != TEXT("describe_scene")) continue;

        Obj.CurrentCount++;

        // Reject submissions below minimum word count
        if (Obj.MinWordCount > 0 && WordCount < Obj.MinWordCount) continue;

        if (Obj.CurrentCount >= (Obj.RequiredCount > 0 ? Obj.RequiredCount : 1))
        {
            Obj.bCompleted = true;
            UE_LOG(LogTemp, Log, TEXT("[Insimul] Writing objective completed: %s"), *Obj.Id);
        }
    }
}

void UQuestSystem::TrackNpcConversationTurn(const FString& NpcId, const FString& TopicTag, const FString& QuestId)
{
    static const TMap<FString, TArray<FString>> TagToTypes = {
        { TEXT("directions"),    { TEXT("ask_for_directions") } },
        { TEXT("order"),         { TEXT("order_food") } },
        { TEXT("haggle"),        { TEXT("haggle_price") } },
        { TEXT("introduction"),  { TEXT("introduce_self") } },
        { TEXT("friendship"),    { TEXT("build_friendship") } },
    };

    TSet<FString> TargetTypes;
    if (!TopicTag.IsEmpty() && TagToTypes.Contains(TopicTag))
    {
        for (const auto& T : TagToTypes[TopicTag]) TargetTypes.Add(T);
    }
    else
    {
        TargetTypes = { TEXT("ask_for_directions"), TEXT("order_food"), TEXT("haggle_price"), TEXT("introduce_self"), TEXT("build_friendship") };
    }

    for (auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;
        if (!QuestId.IsEmpty() && Obj.QuestId != QuestId) continue;
        if (!TargetTypes.Contains(Obj.Type)) continue;
        if (!Obj.NpcId.IsEmpty() && Obj.NpcId != NpcId) continue;

        Obj.CurrentCount++;
        if (Obj.CurrentCount >= (Obj.RequiredCount > 0 ? Obj.RequiredCount : 1))
        {
            Obj.bCompleted = true;
            UE_LOG(LogTemp, Log, TEXT("[Insimul] NPC conversation turn objective completed: %s"), *Obj.Id);
        }
    }
}

void UQuestSystem::TrackConversationInitiation(const FString& NpcId, bool bAccepted, const FString& QuestId)
{
    if (!bAccepted) return;

    for (auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;
        if (!QuestId.IsEmpty() && Obj.QuestId != QuestId) continue;
        if (Obj.Type != TEXT("conversation_initiation")) continue;
        if (!Obj.NpcId.IsEmpty() && Obj.NpcId != NpcId) continue;

        Obj.CurrentCount++;
        if (Obj.CurrentCount >= (Obj.RequiredCount > 0 ? Obj.RequiredCount : 1))
        {
            Obj.bCompleted = true;
            UE_LOG(LogTemp, Log, TEXT("[Insimul] Conversation initiation objective completed: %s"), *Obj.Id);
        }
    }
}

void UQuestSystem::TrackTeachWord(const FString& NpcId, const FString& Word, const FString& QuestId)
{
    FString LowerWord = Word.ToLower();
    for (auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;
        if (!QuestId.IsEmpty() && Obj.QuestId != QuestId) continue;
        if (Obj.Type != TEXT("teach_vocabulary")) continue;
        if (!Obj.NpcId.IsEmpty() && Obj.NpcId != NpcId) continue;

        if (Obj.WordsTaught.Contains(LowerWord)) continue;

        Obj.WordsTaught.Add(LowerWord);
        Obj.CurrentCount++;
        if (Obj.CurrentCount >= (Obj.RequiredCount > 0 ? Obj.RequiredCount : 3))
        {
            Obj.bCompleted = true;
            UE_LOG(LogTemp, Log, TEXT("[Insimul] Teach word objective completed: %s"), *Obj.Id);
        }
    }
}

void UQuestSystem::TrackTeachPhrase(const FString& NpcId, const FString& Phrase, const FString& QuestId)
{
    FString LowerPhrase = Phrase.ToLower();
    for (auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;
        if (!QuestId.IsEmpty() && Obj.QuestId != QuestId) continue;
        if (Obj.Type != TEXT("teach_phrase")) continue;
        if (!Obj.NpcId.IsEmpty() && Obj.NpcId != NpcId) continue;

        if (Obj.PhrasesTaught.Contains(LowerPhrase)) continue;

        Obj.PhrasesTaught.Add(LowerPhrase);
        Obj.CurrentCount++;
        if (Obj.CurrentCount >= (Obj.RequiredCount > 0 ? Obj.RequiredCount : 1))
        {
            Obj.bCompleted = true;
            UE_LOG(LogTemp, Log, TEXT("[Insimul] Teach phrase objective completed: %s"), *Obj.Id);
        }
    }
}

void UQuestSystem::CheckDirectionProximity(const FVector& PlayerPos)
{
    for (auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;

        if (Obj.Type == TEXT("follow_directions") || Obj.Type == TEXT("navigate_language"))
        {
            // Direction step proximity checking stub.
            // Game implementation should deserialize direction_steps/navigation_waypoints
            // from quest data and check player distance against each waypoint target.
        }
    }
}

FString UQuestSystem::GetNextScavengerCategory(int32 LastCategoryIndex)
{
    int32 Next = (LastCategoryIndex + 1) % SCAVENGER_CATEGORIES.Num();
    return SCAVENGER_CATEGORIES[Next];
}

void UQuestSystem::SetPointInBuildingCheck(const FPointInBuildingCheck& Check)
{
    PointInBuildingCheck = Check;
}

TArray<FVector> UQuestSystem::GenerateItemPositions(int32 Count) const
{
    TArray<FVector> Positions;
    const float Radius = 3000.f; // 30 m in UE units

    for (int32 i = 0; i < Count; i++)
    {
        float X = 0.f, Z = 0.f;
        for (int32 Attempt = 0; Attempt < 8; Attempt++)
        {
            float Angle = (PI * 2.f * i) / Count + FMath::FRandRange(0.f, 0.5f);
            float Dist = 1000.f + FMath::FRandRange(0.f, Radius);
            X = FMath::Cos(Angle) * Dist;
            Z = FMath::Sin(Angle) * Dist;
            if (!PointInBuildingCheck.IsBound() || !PointInBuildingCheck.Execute(X, Z))
            {
                break;
            }
        }
        Positions.Add(FVector(X, Z, 50.f)); // Slightly above ground
    }
    return Positions;
}

FVector UQuestSystem::GenerateLocationPosition() const
{
    for (int32 Attempt = 0; Attempt < 8; Attempt++)
    {
        float Angle = FMath::FRandRange(0.f, PI * 2.f);
        float Dist = 2000.f + FMath::FRandRange(0.f, 2000.f);
        float X = FMath::Cos(Angle) * Dist;
        float Z = FMath::Sin(Angle) * Dist;
        if (!PointInBuildingCheck.IsBound() || !PointInBuildingCheck.Execute(X, Z))
        {
            return FVector(X, Z, 0.f);
        }
    }
    // Fallback — push farther out
    float Angle = FMath::FRandRange(0.f, PI * 2.f);
    float Dist = 4000.f + FMath::FRandRange(0.f, 1000.f);
    return FVector(FMath::Cos(Angle) * Dist, FMath::Sin(Angle) * Dist, 0.f);
}

TArray<FVector> UQuestSystem::GetCollectibleItemPositions() const
{
    TArray<FVector> Positions;
    for (const auto& Obj : Objectives)
    {
        if (Obj.bCompleted) continue;
        if (Obj.Type != TEXT("collect_item") && Obj.Type != TEXT("identify_object") && Obj.Type != TEXT("find_vocabulary_items")) continue;
        // Use generated positions for uncollected objectives
        TArray<FVector> ItemPositions = GenerateItemPositions(FMath::Max(1, Obj.RequiredCount - Obj.CurrentCount));
        Positions.Append(ItemPositions);
    }
    return Positions;
}

void UQuestSystem::SetMarkerDebugLabel(AActor* Marker, const FString& Label)
{
    if (!Marker) return;
    // Store debug label as an actor tag for hover tooltip / debug display.
    // Replaces floating 3D text labels with lightweight metadata.
    Marker->Tags.AddUnique(*FString::Printf(TEXT("DebugLabel:%s"), *Label));
}
